exonum/src/node/connect_list.rs

```diff
@@ -14,26 +14,48 @@
 
 //! Mapping between peers public keys and IP-addresses.
 
-use std::{collections::BTreeMap, net::SocketAddr};
+use std::{collections::BTreeMap, net::SocketAddr, net::ToSocketAddrs};
 
 use crypto::PublicKey;
 use node::{ConnectInfo, ConnectListConfig};
 
+/// Network address of the peer.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PeerAddress {
+    /// External address of the peer hostname:port.
+    pub address: String,
+    /// Currently used address resolution.
+    resolved: Option<SocketAddr>,
+    /// Backup address resolutions.
+    resolved_cache: Vec<SocketAddr>,
+}
```

Comment:
I'm not sure that we need to keep the socket addresses after the connection is successfully made. We only need one `SocketAddr` to work and others are effectively useless. If the connection fails at some point and we need to reconnect then we'd better retry resolving as well (e.g., to get a different server from a load-balancing DNS). So I believe we can simply store address strings instead of `PeerAddress` structs.

Reply:
The idea of resolved_cache is to try next ip address if previous resolved address from array was wrong. It is used to track which addresses was not tried before. 
Because of the structure of network the failure of resolve propagates like this: 

call to node connect -> resolve -> network connection pool try to establish connection to resolved address -> cannot connect to peer event -> reconnect(goto 1).

That is why information about previous resolves is stored here, so after this cycle we try different ip.